plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,324 @@
+    private final boolean allowUserNotExist;
+    private final LoadingCache<String, Set<String>> groupCache;
+
+    @Inject
+    public LdapGroupProvider(LdapClient client, LdapGroupProviderConfig ldapConfig)
+    {
+        this.client = requireNonNull(client, "client is null");
+        requireNonNull(ldapConfig, "ldapConfig is null");
+
+        this.userBindSearchPatterns = ldapConfig.getUserBindSearchPatterns();
+        this.userBaseDistinguishedName = Optional.ofNullable(ldapConfig.getUserBaseDistinguishedName());
+        this.groupBaseDistinguishedName = Optional.ofNullable(ldapConfig.getGroupBaseDistinguishedName());
+        this.groupNameAttribute = Optional.ofNullable(ldapConfig.getGroupNameAttribute());
+        this.bindDistinguishedName = ldapConfig.getBindDistinguishedName();
+        this.bindPassword = ldapConfig.getBindPassword();
+        this.allowUserNotExist = ldapConfig.isAllowUserNotExist();
+        this.groupMembershipAttribute = Optional.ofNullable(ldapConfig.getGroupMembershipAttribute());
+        this.groupUserMembershipAttribute = Optional.ofNullable(ldapConfig.getGroupUserMembershipAttribute());
+        if (this.groupMembershipAttribute.isEmpty() && this.groupUserMembershipAttribute.isEmpty()) {
+            throw new LdapGroupMembershipException("both ldap.group-membership-attribute and "
+                    + "ldap.group-user-membership-attribute cannot be empty");
+        }
+        if (this.groupMembershipAttribute.isPresent() && this.groupNameAttribute.isEmpty()) {
+            throw new LdapGroupMembershipException("ldap.group-membership-attribute is set "
+                    + "ldap.group-name-attribute cannot be empty");
+        }
+        this.groupCache = buildNonEvictableCacheWithWeakInvalidateAll(
+                CacheBuilder.newBuilder()
+                        .expireAfterWrite(ldapConfig.getLdapCacheTtl().toMillis(), MILLISECONDS),
+                CacheLoader.from(this::getGroupsWithBindDistinguishedName));
+    }
+
+    @VisibleForTesting
+    void invalidateCache()
+    {
+        groupCache.invalidateAll();
+    }
+
+    @Override
+    public Set<String> getGroups(String user)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            return groupCache.getUnchecked(user);
+        }
+        catch (UncheckedExecutionException e) {
+            throw e;
+        }
+    }
+
+    private Set<String> getGroupsWithBindDistinguishedName(String user)
+    {
+        Set<String> groupSet = Collections.emptySet();
+
+        if (containsSpecialCharacters(user)) {
+            throw new AccessDeniedException("Username contains a special LDAP character");
+        }
+
+        try {
+            for (String userBindSearchPattern : userBindSearchPatterns) {
+                String userSearchFilter = replaceUser(userBindSearchPattern, user);
+                if (this.groupUserMembershipAttribute.isPresent()) {
```

Comment:
Is it possible to have an interface for the implementation and based on the config we can inject them during the runtime ? i.e based on `groupUserMembershipAttribute` or not.

Reply:
yeah good idea .. this is a bit clunky .. let me investigate ..

Reply:
resolved with https://github.com/trinodb/trino/pull/17518/commits/4e4c8fe3d8503bada7f94ae026820d6ab03e63d0
PTAL